GVFS/GVFS.Common/Database/Placeholders.cs

```diff
@@ -0,0 +1,199 @@
+                else
+                {
+                    this.AddPartialFolder(data.Path);
+                }
+            }
+            else
+            {
+                this.AddFile(data.Path, data.Sha);
+            }
+        }
+
+        public void AddFile(string path, string sha)
+        {
+            using (GVFSDatabase.IPooledConnection pooled = this.database.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.File, Sha = sha });
+            }
+        }
+
+        public void AddPartialFolder(string path)
+        {
+            using (GVFSDatabase.IPooledConnection pooled = this.database.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PartialFolder });
+            }
+        }
+
+        public void AddExpandedFolder(string path)
+        {
+            using (GVFSDatabase.IPooledConnection pooled = this.database.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.ExpandedFolder });
+            }
+        }
+
+        public void AddPossibleTombstoneFolder(string path)
+        {
+            using (GVFSDatabase.IPooledConnection pooled = this.database.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Insert(command, new PlaceholderData() { Path = path, PathType = PlaceholderData.PlaceholderType.PossibleTombstoneFolder });
+            }
+        }
+
+        public void Remove(string path)
+        {
+            using (GVFSDatabase.IPooledConnection pooled = this.database.GetPooledConnection())
+            using (SqliteCommand command = pooled.Connection.CreateCommand())
+            {
+                Delete(command, path);
+            }
+        }
+
+        private static void Insert(SqliteCommand command, PlaceholderData placeholder)
+        {
+            command.CommandText = $"INSERT OR REPLACE INTO Placeholders (path, pathType, sha) VALUES (@path, @pathType, @sha);";
+            command.Parameters.Add("@path", SqliteType.Text).Value = placeholder.Path;
+            command.Parameters.Add("@pathType", SqliteType.Integer).Value = (byte)placeholder.PathType;
+            if (placeholder.Sha == null)
+            {
+                command.Parameters.Add("@sha", SqliteType.Text).Value = DBNull.Value;
+            }
+            else
+            {
+                command.Parameters.Add("@sha", SqliteType.Text).Value = placeholder.Sha;
+            }
+
+            command.ExecuteNonQuery();

```

Comment:
Have you tried measuring the performance impact of using prepared statements, and if so, what did you find?

Reply:
I did on the `Contains` method that I had and it did help.  The issue here would be that you need to use the same connection that the prepare was ran against.  This would require the class to keep connections around and do locking for using the commands that are prepared which would really add to the complexity so I will look into that after the PR is smoothed out.

Reply:
I also ran some testing using a simple locking strategy with one prepared connection and command and things were slowed down because of the lock.  The next experiment will be to have a pool of prepared connections and commands to remove the need for the lock.